### PR TITLE
Update API methods that take provisionerId/workerType to take taskQueueId

### LINKED ISSUE #3581: update API methods that take provisionerId/workerType to take taskQueueId
This change should keep compatibility with REST API URL paths, but will be a breaking change for client APIs.  For example, `claimWork(provisionerId, workerType, payload) : result` will become `claimWork(taskQueueId, payload) : result`, but keep the same URL format. 

This does not apply to the worker metadata methods in the queue (which do not have the form `../:provisionerId/:workerType/..` and are experimental / deprecated anyway).

Affected API methods:
 * queue.claimWork
 * queue.pendingTasks
 * purgeCache.purgeCache
 * purgeCache.purgeRequests

### PR DESCRIPTION
**Pull request by alopezz**
Github Bug/Issue: Fixes #3581 

This PR will change API methods that take `provisionerId/workerType` to take taskQueueId instead while remaining compatible on the URL level, but breaking client APIs.

Methods to change:

- [x] `queue.claimWork`
- [x] `queue.pendingTasks`
- [x] `purgeCache.purgeCache`
- [x] `purgeCache.purgeRequests`

When running `yarn generate`, I realized that these changes break generic-worker. I tried my best to adapt it to work while trying to change as little as possible. @djmitche  I need some feedback on whether this was the right thing to do, as it will also be necessary when changing the rest of the API methods.

### CODE REVIEW

**Comment by alopezz** [comment]
So it turns out that tests are failing for generic-worker anyway, even after my attempt to fix what the changes in the client library break. At first glance I fail to see how the failing tests are related to what I've changed, and the feedback is not very helpful; also, those tests take ages to run on my machine for some reason (which makes for a very slow feedback cycle).

**Comment by alopezz** [comment]
I figured out what the problem was and the tests now pass, though I wasn't able to run the tests properly locally.

**Comment by djmitche** [comment]
The golang test output is pretty hard to parse, I agree.  I typically narrow it down to a single test and use `go test  . -tags simple -r TheTestThatsFailing`.

**Comment by alopezz** [comment]
As I start to make the changes in purge-cache, I've noticed, looking at the [schemas](https://github.com/taskcluster/taskcluster/blob/155de774726480f6af3acbbe5344c75de2c37c16/services/purge-cache/schemas/v1/purge-cache-requests.yml), that we don't return a `workerPoolId` in this service (cf. what we do now in the queue service with `taskQueueId`). So I have the question, should we?

**Comment by djmitche** [comment]
Probably, but that's also probably a good place to split off a follow-up issue.

**Comment by alopezz** [comment]
This should be close. What's left:

- multiuser generic-worker tests which fail because of something related to chain-of-trust, which seems to be beyond my reach
- Resolving the conflict with the main branch (`clients/client-web/src/clients/Queue.js`)
- @djmitche What should be the changelog like for this one? Is it `--major` `--users`?

**Comment by djmitche** [comment]
Yes, I think this is major since it changes the API in all of the clients (even if it's compatible at the API level).

**Review by imbstack — APPROVED** [review]
1. this was a lot of work!
2. this looks great!
3. thanks :)

File: workers/generic-worker/mocktc/queue.go [hunk 1]
```diff
@@ -229,6 +226,29 @@ func (queue *Queue) CreateTask(taskId string, payload *tcqueue.TaskDefinitionReq
 		payload.Scopes = []string{}
 	}
 
+	// Handle different possible cases with taskQueueId and provisionerId/workerType
+	var provisionerId, workerType, taskQueueId string
+	if payload.ProvisionerID != "" && payload.WorkerType != "" && payload.TaskQueueID != "" {
```
[the remark this answers is not in the document]
Comment (alopezz) [reply workers/generic-worker/mocktc/queue.go:231]: My intention with this was to reproduce the behavior of the actual service as much as possible, which accepts all three identifiers at the same time as long as they're consistent: https://github.com/taskcluster/taskcluster/blob/52a3b85311e3aa361c97ddfc36eaeff331c57591/services/queue/src/api.js#L519

File: services/queue/src/api.js [hunk 2]
```diff
@@ -1685,26 +1685,27 @@ builder.declare({
...
     'Furthermore, we cache the result in memory for 20 seconds. So consumers',
     'should be no means expect this to be an accurate number.',
     'It is, however, a solid estimate of the number of pending tasks.',
   ].join('\n'),
 }, async function(req, res) {
-  let provisionerId = req.params.provisionerId;
-  let workerType = req.params.workerType;
-  let taskQueueId = joinTaskQueueId(provisionerId, workerType);
+  let taskQueueId = req.params.taskQueueId;
+  let { provisionerId, workerType } = splitTaskQueueId(taskQueueId);
```
[the remark this answers is not in the document]
Comment (alopezz) [reply services/queue/src/api.js]: Totally; in general I just try to keep consistency with what I see around the code I'm changing. Fixed it.

**Comment by petemoore** [comment]
Looks like we are hitting this:

```
    artifacts_multiuser_test.go:171: Did not get back expected task definition in chain of trust certificate:
        &tcqueue.TaskDefinitionRequest{Created:tcclient.Time{wall:0x0, ext:63747848248, loc:(*time.Location)(nil)}, Deadline:tcclient.Time{wall:0x0, ext:63747849148, loc:(*time.Location)(nil)}, Dependencies:[]string{}, Expires:tcclient.Time{wall:0x0, ext:63749057848, loc:(*time.Location)(nil)}, Extra:json.RawMessage(nil), Metadata:tcqueue.TaskMetadata{Description:"Test task", Name:"[TC] Generic Worker CI - TestChainOfTrustUpload", Owner:"generic-worker-ci@mozilla.com", Source:"https://github.com/taskcluster/generic-worker"}, Payload:json.RawMessage(nil), Priority:"lowest", ProjectID:"", ProvisionerID:"test-provisioner", Requires:"all-completed", Retries:1, Routes:[]string{}, SchedulerID:"test-scheduler", Scopes:[]string{}, Tags:map[string]string(nil), TaskGroupID:"XeNVBib9TJ-oqWUt0130uA", TaskQueueID:"test-provisioner/test-ahy5k8jfrnwglh14njebvq-a", WorkerType:"test-ahy5k8jfrnwglh14njebvq-a"}
         ** vs **
        &tcqueue.TaskDefinitionRequest{Created:tcclient.Time{wall:0x0, ext:63747848248, loc:(*time.Location)(nil)}, Deadline:tcclient.Time{wall:0x0, ext:63747849148, loc:(*time.Location)(nil)}, Dependencies:[]string{}, Expires:tcclient.Time{wall:0x0, ext:63749057848, loc:(*time.Location)(nil)}, Extra:json.RawMessage(nil), Metadata:tcqueue.TaskMetadata{Description:"Test task", Name:"[TC] Generic Worker CI - TestChainOfTrustUpload", Owner:"generic-worker-ci@mozilla.com", Source:"https://github.com/taskcluster/generic-worker"}, Payload:json.RawMessage(nil), Priority:"lowest", ProjectID:"", ProvisionerID:"test-provisioner", Requires:"all-completed", Retries:1, Routes:[]string{}, SchedulerID:"test-scheduler", Scopes:[]string{}, Tags:map[string]string(nil), TaskGroupID:"XeNVBib9TJ-oqWUt0130uA", TaskQueueID:"", WorkerType:"test-ahy5k8jfrnwglh14njebvq-a"}
    helper_test.go:474: Removing test directory /task_161225117441705/taskcluster/workers/generic-worker/testdata/TestChainOfTrustUpload...
    helper_test.go:488: Mock HTTP services stopped
--- FAIL: TestChainOfTrustUpload (0.87s)
```

**Review by petemoore — CHANGES_REQUESTED** [review]
I think the generic-worker chain of trust test will need to be fixed. It may be worth checking with @escapewindow if firefox chain of trust workflows might be impacted by this change.

**Comment by alopezz** [comment]
> I think the generic-worker chain of trust test will need to be fixed. It may be worth checking with @escapewindow if firefox chain of trust workflows might be impacted by this change.

Any pointers on where to look at to fix that test? I looked for a while and couldn't figure it out on my own; when I asked @djmitche he told me not to worry about it for the moment.

**Comment by escapewindow** [comment]
> I think the generic-worker chain of trust test will need to be fixed. It may be worth checking with @escapewindow if firefox chain of trust workflows might be impacted by this change.

It's quite possible. It looks like `TaskQueueID` differs between the two. Is this because we're testing generic-worker code that knows about `TaskQueueId` against the production `Queue` that doesn't yet?

When we roll this out, will the queue include `TaskQueueId` where it didn't exist before? Or will newer tasks include it and older tasks have empty `TaskQueueId`s?

One solution may be to ignore `TaskQueueId` in CoT verification, or only verify it if it's a non-empty string, until everything we want to verify uses `TaskQueueId`.

**Comment by djmitche** [comment]
For all tasks, `taskQueueId === provisionerId + '/' + workerType`, so verifying one side of that equality is all that's required.  CoT already verifies the RHS.

We've already added `projectId to tasks, as well.  That one doesn't have an equivalent, but for the moment also isn't significant to a task's identity.

I think a useful invariant to have in CoT is that it only verifies top-level properties that it's aware of, and ignores (or perhaps warns on?) top-level properties that is not aware of.  That would allow us to add additional task properties without this concern, and later adapt CoT to verify them.  I don't know of plans to add additional top-level properties right now, as we've kind of tried to squeeze all the current plans into v41, but historically we've added one or two per year.

**Comment by djmitche** [comment]
I checked with @escapewindow, and the approach we agreed on is that we should arrange to have some tests fail somewhere in the TC codebase when new properties are added to a task definition, with a loud comment saying that they should be added to an allow-list in scriptworker.  Scriptworker will just ignore such "allowed" properties if they exist, and will simultaneously file an issue to plan to actually verify those properties.  That will help us avoid surprises, without blocking on possibly-complicated work to verify new properties in scriptworker.

I can't run the failing g-w test locally, since it requires sudo access, and after re-running it in CI it [passed](https://community-tc.services.mozilla.com/tasks/BfMwqb9kTnqknhV-uuMKkg/runs/0/logs/public/logs/live.log).  So I'm pretty confused about what all that means.

It's not clear to me what about this PR would cause this failure.  But, if the failure were due to upgrading the community deployment (against which these tests run) to v40, then I'd expect other PRs to be failing.

@petemoore I'll try to grab you tomorrow morning to debug this with me.

**Comment by petemoore** [comment]
[`TestChainOfTrustUpload`](https://github.com/alopezz/taskcluster/blob/dff08800ca7d2cf1e9388f9f30068e89de55f796/workers/generic-worker/artifacts_multiuser_test.go#L19-L222) creates, submits, and executes a task with the chain of trust feature enabled, and then compares the task definition it finds in the published `public/chain-of-trust.json` artifact with the task definition it used to create the task. It [disregards](https://github.com/alopezz/taskcluster/blob/dff08800ca7d2cf1e9388f9f30068e89de55f796/workers/generic-worker/artifacts_multiuser_test.go#L161-L169) differences due to `payload`, `tags` and `extra`.

The [test failure output](https://github.com/taskcluster/taskcluster/pull/4277#issuecomment-771661411) shows that the task definition in the chain of trust certificate is different to the version of the task it submitted to the mock queue, and this is because the method that creates the test task doesn't include the `taskQueueId`.

I think there are two alternative fixes:
  * The helper that creates the test task could be updated to include `taskQueueId` json property, or
  * The test could be updated to fetch the task definition from the mock queue, rather than assuming the version it submitted did not get mutated

Probably the latter is a better fix than the former. That said, I've made [a commit](https://github.com/taskcluster/taskcluster/commit/a42b4df2c6f4c88ea24dc25f6975a82ec9871664) for the former in a test branch to test the theory. But we should probably go with the second option, since it is more flexible with regard to future task definition mutations.

**Comment by djmitche** [comment]
Thanks Pete!  I think doing _both_ of those fixes would be great, tbh.

@alopezz does that get you unstuck?  Would you like me to make some pushes to the PR to work on this bit, or anything else?

**Comment by alopezz** [comment]
Green :green_circle:! I applied the second fix first (https://github.com/taskcluster/taskcluster/pull/4277/commits/ffa8c3d3e83280be2267ff0289f8b9bfa567d040), getting the tests to pass, and then I applied the first fix on a second commit.

File: services/queue/src/api.js [hunk 3]
```diff
@@ -935,19 +935,22 @@ let sleep20Seconds = () => {
...
   ] },
   input: 'claim-work-request.yml',
   output: 'claim-work-response.yml',
   title: 'Claim Work',
+  params: {
+    taskQueueId: /^[A-Za-z0-9_-]{1,38}\/[A-Za-z0-9_-]{1,38}$/,
+  },
   description: [
-    'Claim pending task(s) for the given `provisionerId`/`workerType` queue.',
+    'Claim pending task(s) for the given `taskQueueId` queue.',
```
Comment (djmitche) [inline services/queue/src/api.js:953]: "..for the given task queue."?

File: services/queue/src/api.js [hunk 4]
```diff
@@ -935,19 +935,22 @@ let sleep20Seconds = () => {
...
-    'queue:claim-work:<provisionerId>/<workerType>',
+    'queue:claim-work:<taskQueueId>',
     'queue:worker-id:<workerGroup>/<workerId>',
   ] },
   input: 'claim-work-request.yml',
   output: 'claim-work-response.yml',
   title: 'Claim Work',
+  params: {
+    taskQueueId: /^[A-Za-z0-9_-]{1,38}\/[A-Za-z0-9_-]{1,38}$/,
+  },
```
Comment (djmitche) [inline services/queue/src/api.js:951]: This can be added to the list of `params` in the call to `new APIBuilder` at the top of this file.

**Review by djmitche — APPROVED** [review]
Verifying the API-level compat:

```
rubin ~ $ curl https://dustin.taskcluster-dev.net/api/queue/v1/pending/built-in/succeed
{
  "provisionerId": "built-in",
  "workerType": "succeed",
  "taskQueueId": "built-in/succeed",
  "pendingTasks": 0
}
rubin ~ $ curl https://dustin.taskcluster-dev.net/api/queue/v1/pending/built-in%2fsucceed
{
  "provisionerId": "built-in",
  "workerType": "succeed",
  "taskQueueId": "built-in/succeed",
  "pendingTasks": 0
}
```

I successfully ran a task in my dev environment using an older worker, so `queue.claimWork` is OK, as is `purgeCache.purgeRequests`.  As for `purgeCache.purgeCache`,

```
rubin ~ $ echo '{"cacheName": "foo"}' | taskcluster api purgeCache purgeCache builtin succeed
```
(using an old API client)

There are a few minor nits below, but this looks great.  If you want to fix those up I'll click "Merge" :)

**Comment by djmitche** [comment]
I filed https://github.com/mozilla-releng/scriptworker/issues/482 to follow up in scriptworker.
